src/signGovernanceVote.c

```diff
@@ -0,0 +1,432 @@
+			ctx->stage = VOTECAST_STAGE_CONFIRM;
+		} else {
+			ctx->stage = VOTECAST_STAGE_CHUNK;
+		}
+		break;
+
+	case VOTECAST_STAGE_CHUNK:
+		ASSERT(ctx->remainingVotecastBytes == 0);
+		ctx->stage = VOTECAST_STAGE_CONFIRM;
+		break;
+
+	case VOTECAST_STAGE_CONFIRM:
+		ctx->stage = VOTECAST_STAGE_WITNESS;
+		break;
+
+	case VOTECAST_STAGE_WITNESS:
+		ctx->stage = VOTECAST_STAGE_NONE;
+		ui_idle(); // we are done
+		break;
+
+	case VOTECAST_STAGE_NONE:
+		// advanceStage() not supposed to be called after votecast processing is finished
+		ASSERT(false);
+
+	default:
+		ASSERT(false);
+
+	}
+
+	TRACE("Advancing governance voting stage to: %d", ctx->stage);
+}
+
+// this is supposed to be called at the beginning of each APDU handler
+static inline void CHECK_STAGE(sign_governance_vote_stage_t expected)
+{
+	TRACE("Checking stage... current one is %d, expected %d", ctx->stage, expected);
+	VALIDATE(ctx->stage == expected, ERR_INVALID_STATE);
+}
+
+// ============================== INIT ==============================
+
+enum {
+	HANDLE_INIT_CONFIRM_START = 100,
+	HANDLE_INIT_VOTE_PLAN_ID,
+	HANDLE_INIT_PROPOSAL_INDEX,
+	HANDLE_INIT_PAYLOAD_TYPE_TAG,
+	HANDLE_INIT_RESPOND,
+	HANDLE_INIT_INVALID,
+};
+
+static void handleInit_ui_runStep()
+{
+	ui_callback_fn_t* this_fn = handleInit_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_INIT_CONFIRM_START) {
+		ui_displayPrompt(
+		        "Start new",
+		        "governance vote?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(HANDLE_INIT_VOTE_PLAN_ID) {
+		ui_displayHexBufferScreen(
+		        "Vote plan id",
+		        ctx->votePlanId, SIZEOF(ctx->votePlanId),
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_INIT_PROPOSAL_INDEX) {
+		ui_displayUint64Screen(
+		        "Proposal index",
+		        ctx->proposalIndex,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_INIT_PAYLOAD_TYPE_TAG) {
+		ui_displayUint64Screen(
+		        "Payload type tag",
+		        ctx->payloadTypeTag,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_INIT_RESPOND) {
+		respondSuccessEmptyMsg();
+		advanceStage();
+	}
+	UI_STEP_END(HANDLE_INIT_INVALID);
+}
+
+__noinline_due_to_stack__
+void signGovernanceVote_handleInitAPDU(
+        const uint8_t* wireDataBuffer, size_t wireDataSize
+)
+{
+	{
+		//sanity checks
+		CHECK_STAGE(VOTECAST_STAGE_INIT);
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		// parse total length of data to sign
+		ctx->remainingVotecastBytes = parse_u4be(&view);
+		TRACE("Remaining votecast bytes = %u", ctx->remainingVotecastBytes);
+		// we need vote plan id, proposal index, payload type tag
+		// and more data of unknown size for the fragment
+		VALIDATE(view_remainingSize(&view) > VOTE_PLAN_ID_SIZE + 1 + 1, ERR_INVALID_DATA);
+
+		// this is only parsed to be shown in the UI, the whole chunk is passed to hash builder
```

Comment:
What about the `fragment size`, `00` byte and `fragment id tag` items? Have we skipped them somewhere? Or are they not being signed?
https://github.com/input-output-hk/chain-libs/pull/836/files#diff-e7f67ca6b6a6f0099c1d94b8b23146de441863a2c9b3e5b9cd45251b6ae450d2R317-R319

Reply:
Yes, the elements you mention are not being signed, see "Transaction data to sign:" in the document you linked.